Multiparameter persistence, C++ part.

DavidLapous opened the pull request:
Cpp part of PR #817.
List of changes : 
 - new `is_multi_parameter` flag in the simplextree's options to deal with filtration values that behave as array-like instead of float-like.
 - Filtration values are now as const ref to prevent array-like copies
   - This implies adding static members for trivial filtrations : `null_value` and `inf_`
 - new member `filtration_mutable` to have a non-const access to the filtration.
 - For multiparameter filtrations, deactivated the `assign` behavior of the insert to prevent unwanted behaviors (and stay 1-critical). The filtration property is no longer maintained.
 - New array-like structure as `Filtration_value`, child of `std::vector` with numpy-like pointwise operations, in `Simplex_tree/multi_filtrations/Finitely_critical_multi_filtrations.h`
 - The `Simplex_tree/Simplex_tree_multi.h` contains the options of a simplextree multi, functions that can convert a 1-parameter simplextree to a multiparameter simplextree and vice versa, aswell as useful functions to push filtration values on a discrete grid (for discrete computations.
 - I also copied the unit tests of the simplextree to apply them to this new structure.

Feel free to move files/functions if they don't follow Gudhi's standard, or commit if its moral changes.

**Edit 20/02/26:**
PR dependencies:
- #1121
- #1122 
- #1176
- #1191 
- #1297

VincentRouvreau commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -924,7 +933,7 @@ class Simplex_tree {
    */
   template<class InputVertexRange = std::initializer_list<Vertex_handle>>
   std::pair<Simplex_handle, bool> insert_simplex_and_subfaces(const InputVertexRange& Nsimplex,
-                    Filtration_value filtration = 0) {
```
github interface doesn't let me comment on the documentation part of this method, but it is said:
```txt
   * If the insertion fails (the simplex is already there), the bool is set to false. If the insertion
   * fails and the simplex already in the complex has a filtration value strictly bigger than 'filtration',
   * we assign this simplex with the new value 'filtration', and set the Simplex_handle field of the
   * output pair to the Simplex_handle of the simplex.
  ```
We should explain the behaviour difference when `SimplexTreeOptions::is_multi_parameter = true`

VincentRouvreau commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -970,8 +979,20 @@ class Simplex_tree {
-      if (filtration(simplex_one) > filt) {
-        assign_filtration(simplex_one, filt);
+      if (!(filtration(simplex_one) <= filt)) { 
+        if constexpr (SimplexTreeOptions::is_multi_parameter){
```
If we do nothing here when `SimplexTreeOptions::is_multi_parameter = true`, we could avoid a lot of code here.
I think we could test `if constexpr (!SimplexTreeOptions::is_multi_parameter)` upper in the method

VincentRouvreau commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -2246,6 +2277,18 @@ class Simplex_tree {
+//MULTIPERS STUFF
+public: 
+  void set_number_of_parameters(int num){
+		number_of_parameters_ = num;
```
This is a comment for all your changes, but we are using 2 spaces as indentation for the c++ part, cf. https://github.com/GUDHI/gudhi-devel/blob/master/.github/code_conventions.md#code-style

VincentRouvreau commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -1610,12 +1631,22 @@ class Simplex_tree {
+        // in that case, we assume that Filtration_value has a `push_to` member to handle this.
+        max_filt_border_value = Filtration_value(this->number_of_parameters_);
+        for (auto &face_sh : boundary){
+          max_filt_border_value.push_to(filtration(face_sh)); // pushes the value of max_filt_border_value to reach simplex' filtration
```
Put the comment upper the code as lines should not exceed 120 characters per line (when possible), cf. https://github.com/GUDHI/gudhi-devel/blob/master/.github/code_conventions.md#code-style

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,161 @@
+namespace Gudhi::multiparameter::multi_filtrations{
+
+template<typename T=float>
+class Finitely_critical_multi_filtration : public std::vector<T> {
```
Yes, furthermore they are not multi-critical for the moment, so `Multi_filtration_value` seems ok

DavidLapous replied:
Oh no my bad, this can be confusing because multiparameter filtration values can be
 - finitely critical or
 - not finitely critical
This class is meant for the first kind, although it is only compatible with 1-critical filtrations for the moment (i.e. a simplex can only appear at one spot only).

DavidLapous replied:
I've made an doc addition to the FinitelyCriticalFiltrations class to mention this

hschreiber commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,229 @@
+ *    Modification(s):
+ *      - YYYY/MM Author: Description of the modification
+ */
+#ifndef SIMPLEX_TREE_MULTI_H_
```
`Simplex_tree_multi_utilities.h` seems more precise and less misleading to me than `Simplex_tree_multi.h`. What do you think?

DavidLapous replied:
I'm not a big fan of the `utilities` name, but the name can be changed indeed. Let's fix the other issues and we'll decide that once the content of this file is fixed.

hschreiber commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,229 @@
+			simplex.push_back(vertex);
+
+		if (num_parameters > 0)
+		f[0] = st.filtration(simplex_handle);
```
This line needs an indentation for the `if`. The whole file could also be reindented to follow Gudhi's guideline?

hschreiber commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,229 @@
+		std::vector<int> simplex;
+		for (auto vertex : st_multi.simplex_vertex_range(simplex_handle))
+			simplex.push_back(vertex);
+		typename simplextree_multi::Options::value_type f = dimension >= 0 ? st_multi.filtration(simplex_handle)[dimension] : 0;
```
What is the use of `dimension` to possibly be negative?

DavidLapous replied:
put every filtration values of `st` to 0. May be useful to copy a simplextree when we don't care about filtration values.

hschreiber replied:
Ok, I see. You should say this in the description of `dimension`, now that you added the doc.

VincentRouvreau commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -151,16 +151,18 @@ class Simplex_tree {
     void assign_filtration(Filtration_value GUDHI_CHECK_code(f)) { GUDHI_CHECK(f == 0, "filtration value specified for a complex that does not store them"); }
-    Filtration_value filtration() const { return 0; }
+    const Filtration_value& filtration() const  { return null_value; }
+    static constexpr Filtration_value null_value={};
```
If we say:
```cpp
struct MyOptions : Simplex_tree_options_multidimensional_filtration {
  // is_multi_parameter but not store_filtration
  static const bool store_filtration = false;
};

Simplex_tree<MyOptions> stree;
```
We are facing the error:
```
gudhi-devel/src/Simplex_tree/include/gudhi/Simplex_tree.h:165:39: error: the type 'const Filtration_value' {aka 'const Gudhi::multiparameter::multi_filtrations::Finitely_critical_multi_filtration<float>'} of 'constexpr' variable 'Gudhi::Simplex_tree<MyOptions>::Filtration_simplex_base_dummy::null_value' is not literal
     static constexpr Filtration_value null_value={};
                                       ^~~~~~~~~~
```

But maybe this error is *normal* ? I was wondering to see if we should static_assert some error in this case.

hschreiber replied:
A `static_assert` could be a good idea, as the multi parameter tree only differentiate it self from the standard with its filtration values. If `store_filtration == false`, `is_multi_parameter` set to true makes no sense. But the problem could also be solved by removing `null_value` (see Issue #977), then we can just document that `is_multi_parameter` is ignored if `store_filtration == false`.

DavidLapous replied:
A simple fix is to do a `static_assert(!store_filtration && is_multi_parameter)`. thoughts ?

hschreiber commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,229 @@
+
+
+
+// Turns a multi-parameter simplextree into a 1-parameter simplextree
```
Should precise that the filtration values are not token over (which makes sense, but still). Could also add, that for the filtration values, `linear_projection` can be used.

DavidLapous replied:
I'm not sure what you mean, each flatten function take into account filtration values.

hschreiber replied:
Huh? You are right, I think I completely missed a line here as I had the feeling that the filtration value was always put to 0. Which would also explain why I didn't understood the role of the dimension.

hschreiber commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,157 @@
+    return std::vector<Finitely_critical_multi_filtration<T>>(to_convert.begin(), to_convert.end());
+    ;
+  }
+  void push_to(const Finitely_critical_multi_filtration<T>& x) {
```
Once you do the description, you could add that this method is mandatory for `Filtration_value` in the simplex tree.

DavidLapous replied:
I've added that to the concept

hschreiber commented on src/Simplex_tree/example/simplex_tree_multi.cpp:
```diff
@@ -0,0 +1,56 @@
+#include <initializer_list>
+
+struct ST_MULTI {
+public:
```
```suggestion
```
Just a detail: the `public` here is useless, as all members of a `struct` are by default public.

mglisse replied:
True, but I have seen style guides that advocate being explicit like this. So if David likes having `public:` I wouldn't force him to remove it.

hschreiber commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,279 @@
+ * Maximum number of simplices to compute persistence is <CODE>std::numeric_limits<std::uint32_t>::max()</CODE>
+ * (about 4 billions of simplices). */
+struct Simplex_tree_options_multidimensional_filtration {
+public:
```
```suggestion
```
Same as other comment from the example file.

hschreiber commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,279 @@
+		for (auto vertex : st.simplex_vertex_range(simplex_handle))
+			simplex.push_back(vertex);
+
+		if (num_parameters > 0)
+		f[0] = st.filtration(simplex_handle);
```
```suggestion
		if (num_parameters > 0)
		        f[0] = st.filtration(simplex_handle);
```
This line needs an indentation for the if. The whole file could also be reindented to follow Gudhi's guideline?

hschreiber commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,279 @@
+	Simplex_tree<options_multi> &st_multi = *(Gudhi::Simplex_tree<options_multi>*)(splxptr);
+	auto num_parameters = static_cast<unsigned int>(st_multi.get_number_of_parameters());
+	if (grid.size() != num_parameters){
+		std::cerr << "Bad grid !" << std::endl;
+		throw;
```
```suggestion
		throw std::invalid_argument("Bad grid ! Its size is not coherent.");
```

hschreiber commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,279 @@
+				simplex_filtration[parameter] = grid[parameter][simplex_filtration[parameter]];
+		}
+	}
+	return;
```
```suggestion
```
The return here is useless.

hschreiber commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,279 @@
+	std::vector<int> degree_index(degrees.size());
+	int count = 0;
+	for (auto degree : degrees){
+		degree_index[degree] = count; count++;
```
```suggestion
		degree_index[degree] = count++;
```

hschreiber commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,279 @@
+
+}
+
+}	// namespace Gudhi
```
```suggestion
}	// namespace Gudhi::multiparameter
```

hschreiber commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,166 @@
+
+  static std::vector<Finitely_critical_multi_filtration<T>> from_python(const std::vector<std::vector<T>>& to_convert) {
+    return std::vector<Finitely_critical_multi_filtration<T>>(to_convert.begin(), to_convert.end());
+    ;
```
```suggestion
```

VincentRouvreau commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -1610,12 +1624,21 @@ class Simplex_tree {
-                                                              });
+      Filtration_value max_filt_border_value;
+      if constexpr (SimplexTreeOptions::is_multi_parameter) {
+        // in that case, we assume that Filtration_value has a `push_to` member to handle this.
```
This should also be documented in the concept file `gudhi-devel/src/Simplex_tree/concept/FiltrationValue.h`

DavidLapous replied:
I've added a few lines in the concept file.

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -2246,6 +2269,24 @@ class Simplex_tree {
+  /**
+   * \brief Sets the number of parameters of the filtrations if SimplexTreeOptions::is_multi_parameter. 
+   * */
+  void set_number_of_parameters(int num) { number_of_parameters_ = num; }
```
Returning 1 seems more natural to me. But as you wish.

VincentRouvreau commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -151,16 +151,18 @@ class Simplex_tree {
+    Filtration_simplex_base_real() : filt_{} {}
+    void assign_filtration(const Filtration_value& f) { filt_ = f; }
+    const Filtration_value& filtration() const { return filt_; }
+    Filtration_value& filtration() { return filt_; }
```
TBD, but I am not sure we should add this method.
And if we add it we should add it to the *dummy* one.

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,229 @@
+/// @param x 
+/// @param grid 
+/// @return 
+template<typename out_type=int, typename vector_like>
```
It is legal and can have good use cases, but here where out_type is not used anywhere, I don't really see the point.

VincentRouvreau commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,258 @@
+namespace std {
+
+template <>
+class numeric_limits<Gudhi::multiparameter::multi_filtration_type> {
```
I am not sure this code is doing what it is suppose to do:
```cpp
  using Multi_Filtration_values = multi_filtrations::Finitely_critical_multi_filtration<float>;
  std::clog << std::numeric_limits<Multi_Filtration_values>::quiet_NaN() << std::endl;
  // displays: [ ]
  std::clog << std::numeric_limits<Multi_Filtration_values>::infinity() << std::endl;
  // displays: [ ]
```

DavidLapous replied:
I changed the position of this part of the code to the Finitely critical filtration file, which is more logical, and retested on my laptop and I get, as expected `[nan]` and `[inf]`. Does this fix works for you ?

VincentRouvreau replied:
Yes !

VincentRouvreau commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,258 @@
+}
+
+/**
+ * \brief Given a point an a multiparameter discrete grid, pushes the point onto this grid.
```
Typo
```suggestion
 * \brief Given a point on a multiparameter discrete grid, pushes the point onto this grid.
```

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -604,18 +606,25 @@ class Simplex_tree {
    */
-  static Filtration_value filtration(Simplex_handle sh) {
+  static const Filtration_value& filtration(Simplex_handle sh){
     if (sh != null_simplex()) {
       return sh->second.filtration();
     } else {
-      return std::numeric_limits<Filtration_value>::infinity();
+      return inf_;
+    }
+  }
+  static Filtration_value& filtration_mutable(Simplex_handle sh){
+    if (sh != null_simplex()) {
+      return sh->second.filtration();
+    } else {
+      return inf_;
```
I don't have a fixed point of view on this. So its only a matter of what's the gudhi style of accessing a filtration value by reference.

VincentRouvreau commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -2246,6 +2269,34 @@ class Simplex_tree {
+    : std::numeric_limits<Filtration_value>::max(); /**< Default infinite value. */
+
+ private:
+  int number_of_parameters_; /**< Number of parameters of the multi-filtrations when SimplexTreeOptions::is_multi_parameter.-*/
```
`number_of_parameters_` default value is 0. TBD, but wouldn't it be clearer from a user point of view to set it to 2 by default ?

DavidLapous replied:
Yes, why not, as the get number of parameters will always be  1 for non-multiparam simplextrees, this default value seems good.

mglisse replied:
OTOH, a "broken" value like 0 (or even -1) forces the user to be explicit (which can be seen as good or bad).

VincentRouvreau commented on src/Simplex_tree/concept/SimplexTreeOptions.h:
```diff
@@ -31,5 +31,7 @@ struct SimplexTreeOptions {
   static const bool link_nodes_by_label;
   /// If true, Simplex_handle will not be invalidated after insertions or removals.
   static const bool stable_simplex_handles;
+  /// If true, assumes that Filtration_value is vector-like instead of float-like. This also assumes that Filtration_values is a class, which has a push_to method to push a filtration value $x$ onto $this>=0$. 
```
For a better documentation rendering:
```suggestion
  /// If true, assumes that Filtration_value is vector-like instead of float-like. This also assumes that SimplexTreeOptions::Filtration_value is a class, which has a `push_to` method to push a filtration value \f$x\f$ onto \f$this>=0\f$. 
```
![image](https://github.com/GUDHI/gudhi-devel/assets/10407034/2f234aeb-6ac9-4b47-ac1b-61d8237f9ad0)

VincentRouvreau commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,244 @@
+
+    if (num_parameters > 0) f[0] = st.filtration(simplex_handle);
+    st_multi.insert_simplex(simplex, f);
+  }
```
We could take the opportunity to set the number of parameters here, no ?
(My suggestion may be not optimal, not tested)
```suggestion
  }
  simplex.set_number_of_parameters(std::min(static_cast<unsigned int>(default_values.size()), static_cast<unsigned int>(num_parameters - 1)));
```

DavidLapous replied:
Good idea. Note that the filtration that we insert is always of size `num_parameters`, the default values are only meant to optionally initialize the values of this filtration f. So the number of parameter of this simplextree is indeed `num_parameters`.

DavidLapous replied:
I'll also add a test for these functions.

DavidLapous commented:
I'm not sure why the ci Tangential complex test fails on git, I cannot reproduce on my laptop & it doesn't fail on the other pipelines

DavidLapous commented:
@VincentRouvreau I've added your tests to the end of the test file, for the NaN and I had add some checks in the multi filtration file so that the behavior of NaN and inf so that inf and nan filtration values have a proper behavior
 - NaN comparison is always false
 - The push_to method ignores NaN values
 - inf values behave as expected.

Furthermore, thanks to your test (Merci !! 😉) I noticed that there was a bug on the `make_filtration_non_decreasing` in the multiparameter case, which sometimes allows non-increasing values of simplices, that's why I had to change this line

https://github.com/GUDHI/gudhi-devel/blob/a89ff98f87eb6614cdfe8f3fcd8560319a008d71/src/Simplex_tree/test/simplex_tree_multi_unit_test.cpp#L933

I only changed this line and the last test, I hope it didn't break the spirit or your test

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,251 @@
+
+  for (const auto &simplex_handle : st_multi.complex_simplex_range()) {
+    const auto filtration = st_multi.filtration(simplex_handle);
+    const auto degree = st_multi.dimension(simplex_handle);
```
There is also `for_each_simplex` which gives you the dimension for free, but your code is fine.

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,251 @@
+  for (const auto &simplex_handle : st_multi.complex_simplex_range()) {
+    const auto filtration = st_multi.filtration(simplex_handle);
+    const auto degree = st_multi.dimension(simplex_handle);
+    if (std::find(degrees.begin(), degrees.end(), degree) == degrees.end()) continue;
```
Needing this loop kind of defeats the purpose of having degree_index :disappointed:

DavidLapous replied:
I'm not sure how to achieve the same result without this loop. Is there an iterator that is dimension increasing ?

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,251 @@
+  std::vector<int> degree_index(degrees.size());
+  int count = 0;
+  for (auto degree : degrees) {
+    degree_index[degree] = count++;
```
How do you know that all the elements of `degrees` are smaller than the size of `degrees`? Or was degree_index supposed to compute its size based on st_multi.dimension() (or upper_bound_dimension)?

DavidLapous replied:
Thats a bug on my end yes, the degree index should be of size `st_multi.dimension()+1`.

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -2557,6 +2589,34 @@ class Simplex_tree {
+  /**
+   * \brief Sets the number of parameters of the filtrations if SimplexTreeOptions::is_multi_parameter. 
+   * */
+  void set_number_of_parameters(int num) { 
```
Is it intended that we can change the number of parameters of an existing Simplex_tree, or should we ask for it in the constructor?

DavidLapous replied:
It can be useful to change the filtration of an existing simplextree, e.g., adding a filtration while preserving the old ones. But I'm already asking it at construction time in the python interface.

mglisse reviewed (commented):
A first batch of superficial comments (I don't have enough of a global view yet for deeper comments).

DavidLapous commented on src/Persistent_cohomology/example/plain_homology.cpp:
```diff
@@ -24,6 +24,7 @@ struct MyOptions : Gudhi::Simplex_tree_options_minimal {
   // static const bool store_filtration = false;
   // However the persistence algorithm needs this
   static const bool store_key = true;
+  static const bool is_multi_parameter = false;
```
I had some trouble compiling, so at some point I just added this line everywhere. If you say that it's safe, we can removeit here

DavidLapous commented on src/Simplex_tree/concept/FiltrationValue.h:
```diff
@@ -17,7 +17,19 @@
+   * This function is called when using, e.g. `make_filtration_non_decreasing`, as the filtration of a simplex
+   * has to be greater than the filtration of any of its faces.
+   * */ 
+  void push_to(const FiltrationValue f);
```
`push_forward` ? I haven't though a lot about multicritical filtrations, but my first guess to implement this would be to push all of of the filtration values of `this` into each of the `f` ones, and then remove the redundant ones (for $\le$).
I think doing a concept is a good idea.

mglisse replied:
Is "push" a standard term for posets? All I found is the pushout of 2 functors (some form of colimit), is that what you are using here? If the name matches a common definition in category theory, that can be a good reason to use it, but I don't know which one to look at.
(In my intuition, "push"ing one value does not really reflect the symmetry of the operation. Something like smallest_common_majorant (by analogy with GCD) would look more symmetric.)

DavidLapous replied:
I'm not following a standard term for this, so something like you proposed seems great to me.
The analogy is more with the LCM instead of GCD, but "smallest_common_majorant" seems great to me.
As calling multiple times LCM gives the LCM of a family (here `make_filtration_non_decreasing`), this is very moral.

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -2557,6 +2589,34 @@ class Simplex_tree {
+    : std::numeric_limits<Filtration_value>::max(); /**< Default infinite value. */
+
+ private:
+  std::optional<int> number_of_parameters_; /**< Number of parameters of the multi-filtrations when SimplexTreeOptions::is_multi_parameter.-*/
```
not really, I'm not even sure if the optional saves the `int` space for regular simplextrees anyway

mglisse replied:
It doesn't save space, `optional<int>` takes the size of 2 int (an int and a bool, rounded up).

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,322 @@
+    static constexpr bool has_infinity = true; 
+
+    static Gudhi::multiparameter::multi_filtrations::Finitely_critical_multi_filtration<T> infinity() throw(){
+      return Gudhi::multiparameter::multi_filtrations::Finitely_critical_multi_filtration<T>(1, std::numeric_limits<T>::infinity());
```
The reason is that I need to have filtration values that can have coordinates at infinity while being less than an inf value which doesn't depend on the number of parameters. With this implementation it's also compatible with the 1-parameter case, so I'm fine with it. Also, in the multicritical case, this vector is unaware of the number of parameters.

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Box.h:
```diff
@@ -0,0 +1,139 @@
+
+template <typename T>
+inline bool Box<T>::contains(const point_type& point) const {
+  if (point.size() != bottomCorner_.size()) return false;
```
I'll add a warning there

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Line.h:
```diff
@@ -0,0 +1,85 @@
+  ;
+  for (std::size_t i = 0; i < x.size(); i++) {
+    T dir = this->direction_.size() > i ? direction_[i] : 1;
+    t = std::max(t, x[i] / dir);
```
Yes, a check is missing there. We should be able to add orthogonal lines, and then the filtration value should be either infinity if `x` is below the basepoint of the line, or defined by the other coordinates.

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,251 @@
+ * \tparam simplextree_multi A multi simplextree 
+ * \param st Simplextree to copy 
+ * \param st_multi Multiparameter simplextree container to fill. 
+ * \param default_values If given, this vector is assume to be of size `num_parameters-1` and
```
Yes, at the "first possible moment"

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,251 @@
+ * \param dimension The coordinate of the line to choose as a 1 parameter filtration (they are all equivalent).
+ * */
+template <class simplextree_std, class simplextree_multi>
+void flatten_diag(simplextree_std &st, simplextree_multi &st_multi,
```
True, but I initially needed these function only for the python interface, so i kept it simple.

DavidLapous replied:
I'll change that

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,251 @@
+}
+
+/**
+ * \brief Given a point on a multiparameter discrete grid, pushes the point onto this grid.
```
I'm not aware of a terminology. If I remember correctly rivet uses "coarsening".

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,251 @@
+ * */
+template <typename vector_like>
+inline void find_coordinates(vector_like &x, const multi_filtration_grid &grid) {
+  // TODO: optimize with, e.g., dichotomy
```
I almost never use regular grids, as the filtration values usually never fill the space regularly. I still assume that they are sorted. I'll check `lower_bound`, thanks Marc.

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,251 @@
+ * instead of points in this grid.
+ * */
+template <class simplextree_multi>
+void squeeze_filtration(simplextree_multi& st_multi, const multi_filtration_grid &grid, bool coordinate_values = true) {
```
No, maybe coarsen, like rivet, is better

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_multi.h:
```diff
@@ -0,0 +1,251 @@
+  int count = 0;
+  for (auto degree : degrees) {
+    degree_index[degree] = count++;
+    out[degree_index[degree]].reserve(st_multi.num_simplices());
```
I'm not sure that it's too bad, this function is meant to help generating a grid on which to compute "multipersistence", and the output is destroyed before doing any heavy computation. If there is a better value to put there, I can change this.

DavidLapous commented:
Changes since last time:
 - Finitely_critical_filtrations renamed to One_critical_filtration
 - Added a K critical filtration alongside previous one, for the moment defined as a vector of vector, but it may be better to have it as a contiguous one : if the number of parameter is known, a np.array_split(contiguous_stuff, num_parameters) should do the trick.
 - removed some functions that are not necessary yet

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Line.h:
```diff
@@ -0,0 +1,233 @@
+    : basepoint_(x), direction_(v) {}
+template <typename T>
+inline typename Line<T>::point_type
+Line<T>::push_forward(point_type x) const { // TODO remove copy
```
IIUC, x is interpreted as the base point of an upper-right quadrant (a 1-critical thing), the line has a built-in parametrization, and this function returns the parameter of the (smallest) intersection point between the line and the quadrant? I'd like to see comments like that so I don't need to read the code...
(I assume this is called push_forward because we can start from x and increase ("forward") its coordinates as little as possible to bring it to the line ("project" more than "push"?))

DavidLapous replied:
Yes, sorry, I'll add some comments later; this is exactly it. `project` seems great to me. There is also the `push_back` method that does the same thing but with the opposite quadrant; so `project` won't be enough.

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Line.h:
```diff
@@ -0,0 +1,233 @@
+}
+template <typename T>
+template <typename U>
+inline U Line<T>::push_forward2(const point_type &x) const {
```
We need this version because the point could have integer coordinates, but you want some float type for this?

DavidLapous replied:
Yes! I'm not super sure that I need this, but this doesn't cost anything.

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Line.h:
```diff
@@ -0,0 +1,233 @@
+
+private:
+  point_type basepoint_; // any point on the line
+  point_type direction_; // direction of the line
```
`direction_` is indeed assumed to be positive, although non-positive directions could be dealt with with the zigzag module.

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Line.h:
```diff
@@ -0,0 +1,233 @@
+template <typename T>
+template <typename U>
+inline U Line<T>::push_forward2(const point_type &x) const {
+  constexpr const U inf = std::numeric_limits<U>::infinity(); // This disable it for e.g.
+                                                    // ints, but that's good
```
Indeed, I'll fix that

DavidLapous commented on src/Simplex_tree/include/gudhi/Simplex_tree.h:
```diff
@@ -2557,6 +2589,34 @@ class Simplex_tree {
+      return 1;
+  }
+
+  inline static Filtration_value inf_ = std::numeric_limits<Filtration_value>::has_infinity ? 
```
This is an issue with `filtration_mutable` which has to return a reference, I don't really know what's the best way to deal with this

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,1051 @@
+      stuff.pull_to(x);
+    }
+  } 
+  // cannot be const, as gudhi iterators are not const
```
Not sure what you mean. You are not iterating on a Simplex_tree here.

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,1051 @@
+        if (i == j)
+          continue;
+        if (dominates(multi_filtration_[j], multi_filtration_[i], max_error)) {
+          multi_filtration_[i].clear();
```
Are future calls to `dominates` happy to encounter an empty vector? Or does that somehow not happen?

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,1051 @@
+  }
+
+  /*
+  * Checks if b is cleanable with respect to a
```
(tested as `a >= b`)
In my interpretation as a union of upper-right quadrants, it is the larger point that can be removed, not the smaller one.

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,1051 @@
+        if (dominates(multi_filtration_[j], multi_filtration_[i], max_error)) {
+          multi_filtration_[i].clear();
+        }
+        while (j < multi_filtration_.size() && dominates(multi_filtration_[i], multi_filtration_[j], max_error)) {
+          multi_filtration_[j].clear();
+          i--;
+        }
```
Confusing algorithm, decrementing `i` while testing and operating on `j`.

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,1051 @@
+
+    multi_filtration_.push_back(x);
+  }
+  inline void re_clean(){
```
What is the difference with `clean()`?

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,1051 @@
+  end() {
+    return multi_filtration_.end();
+  }
+  inline typename std::vector<
+      One_critical_filtration<T>>::const_iterator
+  begin() const {
```
Note that the following is equivalent (but your code is fine if you like being explicit)
```suggestion
  auto begin() const {
```

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,1051 @@
+    return multi_filtration_;
+  }
+
+  inline std::vector<std::vector<T>> as_VECTOR() const {
```
as_vector vs as_VECTOR, I understand, but maybe better names?

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,1051 @@
+    return out;
+  }
+
+  inline void _clean(bool keep_inf=true) {
```
3rd "clean" function (seems to have a different purpose).

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,1051 @@
+  inline bool operator<=(const Multi_critical_filtration<T,co> &other) const {
+    //check if this curves is below other's curve 
+    // ie for each guy in this, check if there is a guy in other that dominates him
+    for (std::size_t i = 0u; i < multi_filtration_.size(); ++i) {
+      for (std::size_t j = 0u; j < other.multi_filtration_.size(); ++j) {
+        // i <= j 
+        if (dominates(other.multi_filtration_[j], multi_filtration_[i]))
+          continue;
```
`continue` which loop? I think you want the outer one, but C++ understands the inner one.

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,1051 @@
+    if (self.num_generators() != to_compare.num_generators())
+      return false;
+    auto it = to_compare.begin();
+    for (auto i = 0u; i < self.num_generators(); i++) {
+      if (self[i] != *(it++))
+        return false;
```
2 multi-critical lists with the same points in a different order are considered different?

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,1051 @@
+  * Checks if `this`, seen as a birth curve is under the `other` birth curve,
+  *
+  */
+  inline bool operator<(const Multi_critical_filtration<T,co> &other) const {
+    //check if this curves is below other's curve 
```
You say a < b if the curve of a is below the curve of b **and they don't touch anywhere**?
Usually, I would expect: the curve of a is below the curve of b and there is at least 1 place where they differ (they are not equal), although asking for the 2 curves not to touch at all could have some uses I guess. Can you shed some light on how this will be used?

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h:
```diff
@@ -0,0 +1,1051 @@
+namespace std {
+
+template <typename T>
+class numeric_limits<Gudhi::multiparameter::multi_filtrations::
+                         One_critical_filtration<T>> {
```
Where do you need it (trying to understand if it is also needed for multi-critical)?

mglisse commented on src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Box.h:
```diff
@@ -0,0 +1,182 @@
+ * @brief Simple box in \f$\mathbb R^n\f$ .
+ */
+
+namespace Gudhi::multiparameter::multi_filtrations {
```
Can you explain how you choose what goes into `Gudhi::multiparameter::multi_filtrations` vs what goes directly in `Gudhi::multiparameter`?

mglisse commented on src/Multi_filtration/include/gudhi/One_critical_filtration.h:
```diff
@@ -0,0 +1,1378 @@
+   *
+   * @param n Number of parameters.
+   */
+  One_critical_filtration(int n) : Base(n, -T_inf) {};
```
This (and some others below?) should probably be marked `explicit`. 
Would it cause problems to merge it with the next constructor?

hschreiber committed:
doc

hschreiber committed:
Degree_rips_bifiltration

hschreiber committed:
Degree_rips_bifiltration unitary tests

hschreiber committed:
Degree_rips_bifiltration unitary tests

hschreiber committed:
doc

hschreiber committed:
doc

hschreiber committed:
doc

hschreiber committed:
doc

hschreiber committed:
add converters between classes

hschreiber committed:
merge simple_mdspan branche

hschreiber committed:
small fix

hschreiber committed:
forgotten delete after new in tests

hschreiber committed:
forgotten delete after new in tests

hschreiber committed:
Windows fix?

hschreiber committed:
Windows fix?

hschreiber committed:
char* to void* for comparision

hschreiber committed:
Merge remote-tracking branch 'origin/simple_mdspan' into multi_filtrations

hschreiber committed:
upstream merge

hschreiber committed:
number of parameters for simplex tree

hschreiber committed:
Box & Point

hschreiber committed:
line

hschreiber committed:
Merge branch 'GUDHI:master' into multi_persistence

hschreiber committed:
Merge branch 'multi_persistence' of github.com:hschreiber/gudhi-devel into multi_persistence

hschreiber committed:
Merge branch 'GUDHI:master' into multi_filtrations

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
multify/flatten helpers simplified + renamed

hschreiber committed:
add lex comparator

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
Merge branch 'GUDHI:master' into multi_filtrations

hschreiber committed:
start Truc

hschreiber committed:
Merge branch 'GUDHI:master' into multi_persistence

hschreiber committed:
add colex comparator

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
forgotten forward declarations

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
truc part2

hschreiber committed:
more reliable infinity values

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
Truc part3

hschreiber committed:
Merge remote-tracking branch 'origin/representative_cycles' into multi_persistence

hschreiber committed:
Truc part4

hschreiber committed:
Truc part5

hschreiber committed:
Merge branch 'GUDHI:master' into multi_filtrations

hschreiber committed:
operator>> for all multi filtration value types

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
Truc part6

hschreiber committed:
Merge remote-tracking branch 'origin/matrix_changes' into multi_persistence

hschreiber committed:
Truc part6

hschreiber committed:
Truc part7

hschreiber committed:
Slicer io

hschreiber committed:
doc

hschreiber committed:
doc

hschreiber committed:
doc

hschreiber committed:
doc

hschreiber committed:
doc

hschreiber committed:
doc

hschreiber committed:
Merge branch 'GUDHI:master' into multi_persistence

hschreiber committed:
doc

hschreiber committed:
Merge branch 'GUDHI:master' into multi_filtrations

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
Merge remote-tracking branch 'origin/matrix_changes' into multi_persistence

hschreiber committed:
tmp merge 1

hschreiber committed:
Merge remote-tracking branch 'origin/general_cohomology' into multi_persistence

hschreiber committed:
add move constructors to multi filtration value classes

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
cleanup

hschreiber committed:
removal of the useless Co template argument of Multi_parameter_generator

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
bug fix in Dynamic_multi_parameter_filtration

hschreiber committed:
upstream merge

hschreiber committed:
several fixes + re-arrangements

hschreiber committed:
fixes

hschreiber committed:
upstream merge

hschreiber committed:
upstream merge

hschreiber committed:
upstream merge

hschreiber committed:
fixes

hschreiber committed:
Merge remote-tracking branch 'origin/general_cohomology' into multi_persistence

hschreiber committed:
add example from david's branch

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
first update

hschreiber committed:
second update

hschreiber committed:
third update

hschreiber committed:
fourth update

hschreiber committed:
conversion methods

hschreiber committed:
windows fix?

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
more general minus infinity value

hschreiber committed:
upstream merge

hschreiber committed:
windows fix?

hschreiber committed:
typo

hschreiber committed:
Merge branch 'GUDHI:master' into multi_filtrations

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
Merge remote-tracking branch 'hschreiber/multi_persistence' into Multiparameter-Simplextree-C++-part

hschreiber committed:
most persistent cycle

hschreiber committed:
more general build_complex_from_simplex_tree

hschreiber committed:
Merge branch 'GUDHI:master' into multi_filtrations

hschreiber committed:
inverts first and second parameter for Degree_rips_filtration

hschreiber committed:
changes push_to and pull_to of Degree_rips_filtration to be more consistent with the other filtration classes

hschreiber committed:
Merge branch 'GUDHI:master' into multi_filtrations

hschreiber committed:
better bounds for factorize_* in Degree_rips_filtration

hschreiber committed:
Merge branch 'GUDHI:master' into multi_filtrations

hschreiber committed:
Merge remote-tracking branch 'origin/matrix_fix' into vinyard_interface

hschreiber committed:
Merge remote-tracking branch 'origin/matrix_fix' into vinyard_interface

hschreiber committed:
first vineyard batch

hschreiber committed:
vineyard python interface

hschreiber committed:
vineyard rep cycles interface

hschreiber committed:
Merge branch 'GUDHI:master' into vinyard_interface

hschreiber committed:
add possibility of threshold for plotting in vineyard module

hschreiber committed:
Merge branch 'GUDHI:master' into vinyard_interface

hschreiber committed:
small corrections + method to test vineyard validity

hschreiber committed:
threshold for get vineyard methods

hschreiber committed:
revisit color in vineyard plotting + thresholds for getters

hschreiber committed:
Merge branch 'GUDHI:master' into vinyard_interface

hschreiber committed:
vineyard python doc

hschreiber committed:
match cycles with bars in Vineyard

hschreiber committed:
Merge branch 'GUDHI:master' into vinyard_interface

hschreiber committed:
Merge branch 'GUDHI:master' into multi_filtrations

hschreiber committed:
change of behaviour for project_onto_grid

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
upstream merge

hschreiber committed:
add copy constructor for Slicer with different templates

hschreiber committed:
small fixes

hschreiber committed:
Merge remote-tracking branch 'hschreiber/multi_persistence' into Multiparameter-Simplextree-C++-part

hschreiber committed:
Merge branch 'GUDHI:master' into multi_persistence

hschreiber committed:
parallel_for for push_to and coarsen_on_grid

hschreiber committed:
Merge branch 'GUDHI:master' into vinyard_interface

hschreiber committed:
upstream merge

hschreiber committed:
rename vineyard_update

hschreiber committed:
enable cubical complex for vineyard

hschreiber committed:
Merge remote-tracking branch 'origin/vinyard_interface' into multi_persistence

hschreiber committed:
upstream merge

hschreiber committed:
Merge remote-tracking branch 'origin/matrix_changes' into vinyard_interface

hschreiber committed:
Merge remote-tracking branch 'origin/vinyard_interface' into multi_persistence

hschreiber committed:
factorization with new vineyard module

hschreiber committed:
adapted to multi persistence module

hschreiber committed:
Merge remote-tracking branch 'origin/vinyard_interface' into multi_persistence

hschreiber committed:
Merge branch 'GUDHI:master' into multi_filtrations

hschreiber committed:
parallelize some multi filtration methods

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
micro warning fix

hschreiber committed:
copyright year correction

hschreiber committed:
copyright year correction

hschreiber committed:
doc

hschreiber committed:
Update vineyard_base.h cycle type

hschreiber committed:
maybe_unused

hschreiber committed:
Merge branch 'matrix_changes' of github.com:hschreiber/gudhi-devel into matrix_changes

hschreiber committed:
fix ru rep cycles parallel

hschreiber committed:
generalize insert_maximal_cell + compute_dynamic_persistence_from_file + fixes

hschreiber committed:
better template for compute_dynamic_persistence_from_file

hschreiber committed:
doc

hschreiber committed:
vineyard unit test part1

hschreiber committed:
Merge branch 'GUDHI:master' into vinyard_interface

hschreiber committed:
Merge branch 'vinyard_interface' of github.com:hschreiber/gudhi-devel into vinyard_interface

hschreiber committed:
vineyard unit test part1

hschreiber committed:
vineyard unit test part2

hschreiber committed:
vineyard example + doc

hschreiber committed:
python doc

hschreiber committed:
vineyard py test

hschreiber committed:
vineyard doc

hschreiber committed:
missing examples in list

hschreiber committed:
Merge branch 'GUDHI:master' into matrix_changes

hschreiber committed:
split for PR

hschreiber committed:
Merge remote-tracking branch 'origin/master' into multi_filtrations

hschreiber committed:
small fixes in Degree_rips

hschreiber committed:
copyright fix

hschreiber committed:
copyright fix

hschreiber committed:
upstream merge

hschreiber committed:
copyright correction

hschreiber committed:
Merge remote-tracking branch 'origin/vinyard_interface' into multi_persistence

hschreiber committed:
Merge remote-tracking branch 'hschreiber/multi_persistence' into Multiparameter-Simplextree-C++-part

hschreiber committed:
Merge remote-tracking branch 'origin/matrix_insert' into vinyard_interface

hschreiber committed:
review comment fixes

hschreiber committed:
Merge branch 'GUDHI:master' into vinyard_interface

hschreiber committed:
add add_gudhi_py_test_with_file_path to fix vineyard test

hschreiber committed:
add add_gudhi_py_test_with_file_path to fix vineyard test

hschreiber committed:
py vineyard test fix

hschreiber committed:
py vineyard test fix

hschreiber committed:
Merge branch 'GUDHI:master' into vinyard_interface

hschreiber committed:
replacing match syntax for python 3.9

hschreiber committed:
more consistent renaming

hschreiber committed:
typing fix for python 3.9

hschreiber committed:
typing fix for python 3.9

hschreiber committed:
doc

hschreiber committed:
formatting

hschreiber committed:
path manipulation fix for python 3.9

hschreiber committed:
add as_type version

hschreiber committed:
Merge branch 'GUDHI:master' into multi_filtrations

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
small adds

hschreiber committed:
Merge branch 'GUDHI:master' into multi_persistence

hschreiber committed:
Merge branch 'GUDHI:master' into vinyard_interface

hschreiber committed:
Merge branch 'GUDHI:master' into multi_persistence

hschreiber committed:
Merge branch 'GUDHI:master' into vinyard_interface

hschreiber committed:
multipers update

hschreiber committed:
Merge remote-tracking branch 'origin/vinyard_interface' into multi_persistence

hschreiber committed:
chatty optimization for unordered set column

DavidLapous committed:
fix underflow

hschreiber committed:
complex scc write-read fix

hschreiber committed:
Merge pull request #1 from DavidLapous/patch-1

fix underflow

hschreiber committed:
doc

hschreiber committed:
upstream merge

hschreiber committed:
Merge branch 'vinyard_interface' of github.com:hschreiber/gudhi-devel into vinyard_interface

hschreiber committed:
doc

hschreiber committed:
vineyard python tests redone part1

hschreiber committed:
vineyard python tests redone part2

hschreiber committed:
Merge branch 'GUDHI:master' into vinyard_interface

hschreiber committed:
separating vineyard classes from plotting

hschreiber committed:
generalization of vineyard plot options

hschreiber committed:
vineyard plot better handling of inf vines

hschreiber committed:
vineyard plot better handling of inf vines

hschreiber committed:
more arguments for vine plotting

hschreiber committed:
Merge remote-tracking branch 'origin/vinyard_interface' into multi_persistence

hschreiber committed:
Merge branch 'multi_persistence' of github.com:hschreiber/gudhi-devel into multi_persistence

hschreiber committed:
multi filtration project to grid fix

hschreiber committed:
doc update for filtration_simplex_range

hschreiber committed:
simplex tree operator<< and >> as friend

hschreiber committed:
upstream merge

hschreiber committed:
replacing filtration_simplex_range with for_each_simplex in operator<< of simplex tree

hschreiber committed:
CI fix ?

hschreiber committed:
review comments for multi simplex tree example

hschreiber committed:
CI fix ?

hschreiber committed:
factorization + Windows fix

hschreiber committed:
factorization + Windows fix

hschreiber committed:
generalization of module helpers input and output

hschreiber committed:
Merge branch 'GUDHI:master' into multi_filtrations

hschreiber committed:
compute_coordinates_in_grid template order

hschreiber committed:
upstream merge

hschreiber committed:
Merge remote-tracking branch 'origin/multi_filtrations' into multi_persistence

hschreiber committed:
Merge branch 'multi_persistence' of github.com:hschreiber/gudhi-devel into multi_persistence

hschreiber committed:
contiguous numpy array condition

hschreiber committed:
Doc + set_slicer

hschreiber committed:
tiny simplification

hschreiber committed:
doc

hschreiber committed:
summand unit tests

hschreiber committed:
summand unit tests

hschreiber committed:
unit tests + removing box from module

hschreiber committed:
module unit tests

hschreiber committed:
module unit tests

hschreiber committed:
module unit tests

hschreiber committed:
add bars on batch of lines for module

hschreiber committed:
typo

hschreiber committed:
move summand flat methods to python binding

hschreiber committed:
add num_parameter to simplex tree

hschreiber committed:
More precise check for dynamic multi filtrations

hschreiber committed:
upstream merge

hschreiber committed:
upstream merge

hschreiber committed:
Merge remote-tracking branch 'hschreiber/multi_persistence' into Multiparameter-Simplextree-C++-part

hschreiber committed:
upstream merge

hschreiber committed:
remove unnecessary nanobind::any_contig

hschreiber committed:
Merge remote-tracking branch 'hschreiber/multi_persistence' into Multiparameter-Simplextree-C++-part
